README
Data-MusicXML

This is an experimental framework for parsing MusicXML files.

If you are in this sort of business, this may contain interesting
information.

No ready to use tools and modules yet.

COPYRIGHT AND LICENCE

Copyright (C) 2016 Johan Vromans

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Data-MusicXML

0.03    2016-12-04

	Actualize parse script.

0.02    2016-12-04

        Reworked from XML::Simple to XML::LibXML.
	Proof-of-concept 'Yellow Dog Blues' works.

0.01    2016-12-01

	First version, not released on an unsuspecting world.



Makefile.PL
#! perl

# Ease the life of the CPAN testers.
exit 0 if $ENV{AUTOMATED_TESTING} && $] < 5.010000;

require 5.010;

use ExtUtils::MakeMaker;

my $master = 'lib/Data/MusicXML.pm';
my $version = MM->parse_version($master);

WriteMakefile
(
 NAME		=> 'Data::MusicXML',
 ABSTRACT_FROM  => $master,
 VERSION	=> "$version",
 AUTHOR	        => 'Johan Vromans <jvromans@squirrel.nl>',

 INSTALLDIRS	=> 'site',
 EXE_FILES	=> [ 'script/parse' ],
 PL_FILES	=> {
		    'script/pp.PL' => [ 'script/parse' ]
		   },

 PREREQ_PM      => {
		    'XML::LibXML'  => 2.0121,
		   },
 BUILD_REQUIRES  => {
		    'Test::More'   => 1.001014,
		   },

 MIN_PERL_VERSION => '5.010000',
 LICENSE => 'perl',
 META_MERGE => {
     'meta-spec' => { version => 2 },
     resources        => {
	 repository   => {
	     type => 'git',
	     url  => 'git://github.com/sciurius/perl-Data-MusicXML',
	     web  => 'https://github.com/sciurius/perl-Data-MusicXML',
         },
     },
 },

);

1;


lib/Data/MusicXML.pm
#! perl

package Data::MusicXML;

use warnings;
use strict;
use Carp qw( carp croak );

use XML::LibXML;
use DDumper;
use Encode qw( decode_utf8 encode_utf8 );

=head1 NAME

Data::MusicXML - Framework for parsing MusicXML

=cut

our $VERSION = '0.03';

=head1 SYNOPSIS

    use Data::MusicXML;

    my $p = Data::MusicXML->new();
    $p->processfile('Yellow_Dog_Blues.xml');

=cut

sub new {
    my ( $pkg, $options ) = @_;

    my $self = bless( { }, $pkg );

    for ( qw( catalog trace debug verbose ) ) {
	$self->{$_} = $options->{$_} if exists $options->{$_};
    }

    return $self;
}

sub processfiles {
    my ( $self, @files ) = @_;
    $self->{songix} = 0;
    foreach ( @files ) {
	$self->processfile($_);
	$self->{songix}++;
    }
}

# Process a subtree, identified by $key.
sub _process {
    my ( $self, $data, $key, $handler, $ctx ) = @_;

    my @nodes = $data->fn($key);
    warn("No $key nodes found\n"), return unless @nodes;

    my $ix = 0;
    foreach ( @nodes ) {

	# Establish context and link in parent context, if any.
	my $c = $self->{_ctx}->{$key} =
	  { path => join("/", $ctx->{path}, $key) };
	$c->{_parent} = $ctx if $ctx;

	$handler->( $self, $ix+1, $_, $c );

	$ix++;
    }
}

sub processfile {
    my ( $self, $file ) = @_;

    croak( decode_utf8($file) . ": $!" ) unless -r $file;
    my $parser = XML::LibXML->new;
    $parser->load_catalog( $self->{catalog} );
    $parser->set_options( { no_cdata => 1 } );
    my $data = $parser->parse_file($file);

    # <score-partwise>
    #   <work>
    #      <work-title>Yellow Dog Blues</work-title>
    #   </work>
    #   <movement-title>Yellow Dog Blues</movement-title>
    #   <identification>
    #     <encoding>
    #        <software>MuseScore 2.0.3</software>
    #        ...
    #     </encoding>
    #   </identification>
    #   <defaults ... />
    #   <credit page="1" ... />
    #   <credit page="1" ... />
    #   <part-list>
    #      <score-part id="P1" ... />
    #   </part-list>
    #   <part id="P1" ... />
    # </score-partwise>

    # print DDumper($data);

    my $root = "/score-partwise";
    my $rootnode = $data->findnodes($root)->[0];

    my $song = { title => 'NoName',
		 composer => 'NoBody',
		 key => 'C',
		 tempo => 100,
		 index => $self->{songix},
		 parts => [] };

    if ( my $d = $rootnode->fn1('movement-title') ) {
	$song->{title} = $song->{'movement-title'} = $d->to_literal;
	warn( "Title: ", decode_utf8($song->{title}), "\n" )
	  if $self->{debug};
    }
    if ( my $d = $rootnode->fn1('work/work-title') ) {
	$song->{title} = $song->{'work-title'} = $d->to_literal;
	warn( "Title: ", decode_utf8($song->{title}), "\n" )
	  if $self->{debug};
    }

    if ( my $d = $rootnode->fn1('identification/creator[@type=\'composer\']') ) {
	$song->{composer} = $d->to_literal;
	warn( "Composer: ", decode_utf8($song->{composer}), "\n" )
	  if $self->{debug};
    }

    $self->{song} = $song;

    $self->_process( $rootnode, "part", \&process_part,
		     { path => $root } );

    DDumper($song);

    use Data::MusicXML::iRealPro;
    for ( my $ix = 0; $ix < @{ $song->{parts} }; $ix++ ) {
	Data::MusicXML::iRealPro->to_irealpro( $song, $ix );
    }

}

sub process_part {
    my ( $self, $part, $data, $ctx ) = @_;

    # <part id="P1">
    #   <measure ... />
    # </part>

    my $this = {};
    push( @{ $self->{song}->{parts} }, $this );

    $this->{id} = $data->fn1('@id')->to_literal;
    warn( "Part $part: ", decode_utf8($this->{id}), "\n")
      if $self->{debug};

    # print DDumper($data);

    $this->{sections} = [];
    $self->_process( $data, "measure", \&process_measure, $ctx );

}

sub process_measure {
    my ( $self, $measure, $data, $ctx ) = @_;

    # <measure number="24">
    #   <direction ... />
    #   <note ... />
    #   <harmony ... />
    #   <barline ... />
    # </measure>

    use Data::MusicXML::Data qw( @clefs );

    my $this = $self->{song}->{parts}->[-1];

    use feature qw(state);
    state $lastchord;

    my $mark = "";
    foreach ( @{ $data->fn('direction/direction-type/rehearsal') } ) {
	$mark = $_->to_literal;
    }

    if ( $mark ) {
	push( @{ $this->{sections} },
	      { mark => $mark, measures => [] } );
    }
    elsif ( @{ $this->{sections} } == 0 ) {
	$this->{sections} = [ { measures => [] } ];
    }
    $this = $this->{sections}->[-1];

    my $clef = "";
    my $mode = "major";
    foreach ( @{ $data->fn('attributes/key/*') } ) {
	if ( $_->nodeName eq "fifths" ) {
	    $clef = $clefs[$_->to_literal];
	}
	if ( $_->nodeName eq "mode" ) {
	    $mode = $_->to_literal;
	}
    }

    printf STDERR ( "Measure %2d: \"%s\" %s%s %s\n",
		    $measure,
		    $data->fn1('@number')->to_literal,
		    $mark ? "[$mark] " : "",
		    $clef ? ( $clef, $mode ) : ( "", "" ),
		  )
      if $self->{debug};

    $clef .= "-" if $mode eq 'minor';
    $self->{song}->{key} ||= $clef;

    # warn DDumper($data);

    if ( my $d = $data->fn1('sound/tempo') ) {
	$ctx->{tempo} = $d->to_literal;
	$ctx->{_parent}->{tempo} = $ctx->{tempo};
	print STDERR ( " Tempo: ", $ctx->{tempo}, "\n" )
	  if $self->{debug};
    }
    else {
	$ctx->{tempo} = $ctx->{_parent}->{tempo};
    }

    if ( my $d = $data->fn('attributes/time/*') ) {
	foreach ( @$d ) {
	    $ctx->{_parent}->{beats} = $ctx->{beats} = $_->to_literal
	      if $_->nodeName eq "beats";
	    $ctx->{_parent}->{'beat_type'} = $ctx->{'beat_type'} = $_->to_literal
	      if $_->nodeName eq "beat-type";
	}
	print STDERR ( " Beats: ",
		       $ctx->{beats}, "/", $ctx->{'beat_type'},
		       "\n" ) if $self->{debug};
	$this->{time} = $ctx->{beats} ."/". $ctx->{'beat_type'};
    }
    else {
	$ctx->{beats} = $ctx->{_parent}->{beats};
	$ctx->{'beat_type'} = $ctx->{_parent}->{'beat_type'};
    }

    if ( my $d = $data->fn1('attributes/staves') ) {
	$ctx->{staves} = $d->to_literal;
    }

    if ( my $d = $data->fn1('attributes/divisions') ) {
	$ctx->{_parent}->{divisions} =
	$ctx->{divisions} = $d->to_literal;
	print STDERR ( " Divisions: ", $ctx->{divisions}, "\n" )
	  if $self->{debug};
    }
    else {
	$ctx->{divisions} = $ctx->{_parent}->{divisions};
    }

    # Process note and harmony nodes, in order.
    my ( $n, $h );
    $ctx->{currentbeat} = 0;
    my @chords = ( "_" ) x $ctx->{beats};
    foreach ( @{ $data->fn('note | ./harmony') } ) {
	print STDERR ("== beat: ", $ctx->{currentbeat}, "\n" )
	  if $self->{debug};
	$self->process_note( ++$n, $_, $ctx )
	  if $_->nodeName eq "note";
	if ( $_->nodeName eq "harmony" ) {
	    $chords[$ctx->{currentbeat}] =
	      $self->process_harmony( ++$h, $_, $ctx );
	}
    }

    if ( $chords[0] eq "_" && $lastchord ) {
	$chords[0] = $lastchord;
    }
    push( @{ $this->{measures} },
	  { number => $data->fn1('@number')->to_literal,
	    chords => [ @chords ] } );

    pop(@chords) while @chords && $chords[-1] eq "_";
    $lastchord = $chords[-1] if @chords;
}

sub process_note {
    my ( $self, $note, $data, $ctx ) = @_;

    use Data::MusicXML::Data qw( %durations );

    # Duration, in beats.
    my $duration = $data->fn('duration')->[0]->to_literal
      / $ctx->{divisions};
    # Duration is the actual duration, dots included.
    # $duration *= 1.5 if $data->fn('dot');

    my $root;

    if ( my $d = $data->fn1('pitch') ) {
	$root = $d->fn('step')->[0]->to_literal;
	foreach ( @{ $d->fn('alter') } ) {
	    $root .= 'b' if $_->to_literal < 0;
	    $root .= '#' if $_->to_literal > 0;
	}
	if ( my $d = $d->fn1('octave') ) {
	    $root .= $d->to_literal;
	}
    }
    elsif ( $data->fn1('rest') ) {
	$root = 'rest';
    }

    printf STDERR ("Note %3d: %s %s x=%d d=%.2f s=%d\n",
		   $note,
		   $root,
		   $data->fn('type')->[0]->to_literal,
		   eval { $data->fn1('default-x')->to_literal } || 0,
		   $duration,
		   eval { $data->fn1('staff')->to_literal } || 1,
		  )
      if $self->{debug};

    $ctx->{currentbeat} += $duration
      unless $data->fn1('chord');
}

sub process_harmony {
    my ( $self, $harmony, $data, $ctx ) = @_;

#    warn DDumper($data);

    my $root = $data->fn1('root/root-step')->to_literal;
    foreach ( @{ $data->fn('root/root-alter') } ) {
	$root .= 'b' if $_->to_literal < 0;
	$root .= '#' if $_->to_literal > 0;
    }

    my $tquality = "";
    my $quality = $data->fn1('kind')->to_literal;
    if ( my $d = $data->fn1('kind/@text') ) {
	$tquality = $d->to_literal;
    }

    my @d;
    foreach ( @{ $data->fn('degree') } ) {
	push( @d, [ $_->fn1('degree-value')->to_literal,
		    $_->fn1('degree-alter')->to_literal,
		    $_->fn1('degree-type')->to_literal ] );
    }

    printf STDERR ( "Harm %3d: %s%s %s\n",
		    $harmony, $root, $quality, $tquality )
      if $self->{debug};

    return [ $root, $quality, $tquality, @d ? \@d : () ];

}


################ Convenience ################

# Convenient short for subnodes. Returns a nodelist.
sub XML::LibXML::Node::fn {
    $_[0]->findnodes( './' . $_[1] );
}

# Convenient short for single subnode. Returns a node.
sub XML::LibXML::Node::fn1 {
    my $nl = $_[0]->findnodes( './' . $_[1] . '[1]' );
    return unless $nl;
    $nl->[0];
}

=head1 AUTHOR

Johan Vromans, C<< <JV at CPAN dot org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-data-musicxml at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-MusicXML>. I
will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Data::MusicXML

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-MusicXML>

=item * Search CPAN

L<http://search.cpan.org/dist/Data-MusicXML>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2016 Johan Vromans, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Data::MusicXML


lib/Data/MusicXML/Data.pm
#! perl

use strict;
use warnings;

package Data::MusicXML::Data;

our $VERSION = "0.01";

use parent qw( Exporter );

our @EXPORT;
our @EXPORT_OK;

our @clefs =
  ( 'C',				 #  0
    'G', 'D', 'A', 'E', 'B', 'F#', 'C#', #  1 .. 7
    'Gb', 'Db', 'Ab', 'Eb', 'Bb', 'F',	 # -6 .. -1
  );

push( @EXPORT_OK, '@clefs' );

our %durations =
  ( whole    => 4,
    half     => 2,
    quarter  => 1,
    eighth   => 0.5,
    '16th'   => 0.25,
    '32nd'   => 0.125,
  );

push( @EXPORT_OK, '%durations' );

our %note_numbers =
  ( 'C'	  =>  0,
    'C#'  =>  1,
    'Db'  =>  1,
    'D'	  =>  2,
    'D#'  =>  3,
    'Eb'  =>  3,
    'E'	  =>  4,
    'F'	  =>  5,
    'F#'  =>  6,
    'Gb'  =>  6,
    'G'	  =>  7,
    'G#'  =>  8,
    'Ab'  =>  8,
    'A'	  =>  9,
    'A#'  => 10,
    'Bb'  => 10,
    'B'	  => 11,
  );

push( @EXPORT_OK, '%note_numbers' );

# http://usermanuals.musicxml.com/MusicXML/Content/ST-MusicXML-kind-value.htm
our %harmony_kinds =
  (
    # Triads.
    major		  => "",
    minor		  => "",
    augmented		  => "",
    diminished		  => "",

    # Sevenths.
    dominant		  => "",
    'major-seventh'	  => "",
    'minor-seventh'	  => "",
    'diminished-seventh'  => "",
    'augmented-seventh'	  => "",
    'half-diminished'	  => "",
    'major-minor'	  => "",

    # Sixths.
    'major-sixth'	  => "",
    'minor-sixth'	  => "",

    # Ninths.
    'dominant-ninth'	  => "",
    'major-ninth'	  => "",
    'minor-ninth'	  => "",

    # 11ths.
    'dominant-11th'	  => "",
    'major-11th'	  => "",
    'minor-11th'	  => "",

    # 13ths.
    'dominant-13th'	  => "",
    'major-13th'	  => "",
    'minor-13th'	  => "",

    # Suspended.
    'suspended-second'	  => "",
    'suspended-fourth'	  => "",

    # Functional sixths.
    Neapolitan		  => "",
    Italian		  => "",
    French		  => "",
    German		  => "",

    # Other.
    pedal		  => "", # pedal-point bass
    power		  => "", # perfect fifth
    Tristan		  => "",

    # The "other" kind is used when the harmony is entirely composed of
    # add elements.
    other		  => "",

    # The "none" kind is used to explicitly encode absence of chords
    # or functional harmony.
    none		  => "",
  );

push( @EXPORT_OK, '%harmony_kinds' );

1;


lib/Data/MusicXML/iRealPro.pm
#! perl

use strict;
use warnings;
use utf8;
use Carp;

package Data::MusicXML::iRealPro;

our $VERSION = "0.01";

sub to_irealpro {
    my ( $self, $song, $part ) = @_;

    print( "\n" ) if $part+$song->{index};
    print( "Song ", 1+$part+$song->{index}, ": ",
	   $song->{title}, " (", $song->{composer}, ")\n",
	   "Style: Medium Swing;",
	   " key: ", $song->{key}, ";",
	   " tempo: ", $song->{tempo}, ";",
	   " repeat: 1\n",
	   "\n" );

    $part = $song->{parts}->[$part];
    my $irp = "";

    foreach my $s ( @{ $part->{sections} } ) {
	if ( $s->{mark} ) {
	    $irp .= '[*' . $s->{mark};
	}
	else {
	    $irp .= "[";
	}

	if ( $s->{time} ) {
	    $irp .= " T" . timesig( $s->{time} );
	}

	my $i = 0;
	foreach my $m ( @{ $s->{measures} } ) {
	    $irp .= " " . join(" ", map { irpchord($_) } @{ $m->{chords} }) . " |";
	    if ( ( $i +=  @{ $m->{chords} } ) >= 16 ) {
		$irp .= "\n|";
		$i = 0;
	    }
	}
	$irp =~ s/ \|(\n\|)?$//;
        $irp .= " ]\n";
    }

    # Make sure alpha items (chords) are separated by commas if necessary.
    $irp =~ s/([[:alnum:]]+)\s+([[:alnum:]])/$1, $2/g;
    $irp =~ s/([[:alnum:]]+)\s+([[:alnum:]])/$1, $2/g;

    # Add end bar.
    # $irp =~ s/\n$/ Z /;

    print( $irp, "\n" );
}

################ Chords ################

my %harmony_kinds =
  (
    # Triads.
    major		  => "",
    minor		  => "-",
    augmented		  => "+",
    diminished		  => "o",

    # Sevenths.
    dominant		  => "7",
    'major-seventh'	  => "^7",
    'minor-seventh'	  => "-7",
    'diminished-seventh'  => "o7",
    'augmented-seventh'	  => "+7",
    'half-diminished'	  => "h",
    'major-minor'	  => "-^7",

    # Sixths.
    'major-sixth'	  => "6",
    'minor-sixth'	  => "-6",

    # Ninths.
    'dominant-ninth'	  => "9",
    'major-ninth'	  => "^9",
    'minor-ninth'	  => "-9",

    # 11ths.
    'dominant-11th'	  => "11",
    'major-11th'	  => "^11",
    'minor-11th'	  => "-11",

    # 13ths.
    'dominant-13th'	  => "13",
    'major-13th'	  => "^13",
    'minor-13th'	  => "-13",

    # Suspended.
    'suspended-second'	  => "sus2",
    'suspended-fourth'	  => "sus4",

  );

my %chordqual =
  (  ""			=> '',
     "+"		=> 'p',
     "-"		=> 'm',
     "-#5"		=> 'mx5',
     "-11"		=> 'm11',
     "-6"		=> 'm6',
     "-69"		=> 'm69',
     "-7"		=> 'm7',
     "-7b5"		=> 'm7b5',
     "-9"		=> 'm9',
     "-^7"		=> 'mv7',
     "-^9"		=> 'mv9',
     "-b6"		=> 'mb6',
     "11"		=> '11',
     "13"		=> '13',
     "13#11"		=> '13x11',
     "13#9"		=> '13x9',
     "13b9"		=> '13b9',
     "13sus"		=> '13sus',
     "2"		=> '2',
     "5"		=> '5',
     "6"		=> '6',
     "69"		=> '69',
     "7"		=> '7',
     "7#11"		=> '7x11',
     "7#5"		=> '7x5',
     "7#9"		=> '7x9',
     "7#9#11"		=> '7x9x11',
     "7#9#5"		=> '7x9x5',
     "7#9b5"		=> '7x9b5',
     "7alt"		=> '7alt',
     "7b13"		=> '7b13',
     "7b13sus"		=> '7b13sus',
     "7b5"		=> '7b5',
     "7b9"		=> '7b9',
     "7b9#11"		=> '7b9x11',
     "7b9#5"		=> '7b9x5',
     "7b9#9"		=> '7b9x9',
     "7b9b13"		=> '7b9b13',
     "7b9b5"		=> '7b9b5',
     "7b9sus"		=> '7b9sus',
     "7sus"		=> '7sus',
     "7susadd3"		=> '7susadd3',
     "9"		=> '9',
     "9#11"		=> '9x11',
     "9#5"		=> '9x5',
     "9b5"		=> '9b5',
     "9sus"		=> '9sus',
     "^"		=> 'v',
     "^13"		=> 'v13',
     "^7"		=> 'v7',
     "^7#11"		=> 'v7x11',
     "^7#5"		=> 'v7x5',
     "^9"		=> 'v9',
     "^9#11"		=> 'v9x11',
     "add9"		=> 'add9',
     "alt"		=> '7alt',
     "h"		=> 'h',
     "h7"		=> 'h7',
     "h9"		=> 'h9',
     "o"		=> 'o',
     "o7"		=> 'o7',
     "sus"		=> 'sus',
  );

sub irpchord {
    my ( $c ) = @_;
    return $c unless ref($c) eq 'ARRAY';
    my ( $root, $quality, $text, $degree ) = @$c;
    if ( exists $harmony_kinds{$quality} ) {
	$text = $harmony_kinds{$quality};
    }
    else {
	$text = "?";
    }

    $degree ||= [];

    foreach ( @$degree ) {
	my ( $value, $alter, $type ) = @$_;
	next unless $type eq 'add' || $type eq 'alter';
	$text .= 'b' if $alter < 0;
	$text .= '#' if $alter > 0;
	$text .= $value;
    }

    return $root . $text if exists $chordqual{$text};
    return $root . '*' . $text . '*';
}

################ Time Signatures ################

my $_sigs;

sub timesig {
    my ( $time ) = @_;
    $_sigs ||= { "2/2" => "22",
		 "3/2" => "32",
		 "2/4" => "24",
		 "3/4" => "34",
		 "4/4" => "44",
		 "5/4" => "54",
		 "6/4" => "64",
		 "7/4" => "74",
		 "2/8" => "28",
		 "3/8" => "38",
		 "4/8" => "48",
		 "5/8" => "58",
		 "6/8" => "68",
		 "7/8" => "78",
		 "9/8" => "98",
		"12/8" => "12",
	       };

    $_sigs->{ $time }
      || Carp::croak("Invalid time signature: $time");
}

1;


script/parse.pl
#!/usr/bin/perl

# parse -- parse MusicXML song data

# Author          : Johan Vromans
# Created On      : Thu Dec  1 20:15:22 2016
# Last Modified By: Johan Vromans
# Last Modified On: Sun Dec  4 17:14:44 2016
# Update Count    : 7
# Status          : Unknown, Use with caution!

################ Common stuff ################

use strict;
use warnings;
use FindBin;
use lib "$FindBin::Bin/../CPAN";
use lib "$FindBin::Bin/../lib";
use Data::MusicXML;

################ Setup  ################

# Process command line options, config files, and such.
my $options;
$options = app_setup( "parse", $Data::MusicXML::VERSION );

################ Presets ################

$options->{catalog} = "$FindBin::Bin/../res/catalog.xml";
$options->{trace} = 1   if $options->{debug};
$options->{verbose} = 1 if $options->{trace};

################ Activate ################

main($options);

################ The Process ################

sub main {
    my ($options) = @_;
    binmode(STDERR,':utf8');
    Data::MusicXML->new($options)->processfiles(@ARGV);
}

################ Options and Configuration ################

use Getopt::Long 2.13 qw( :config no_ignorecase );
use File::Spec;
use Carp;

# Package name.
my $my_package;
# Program name and version.
my ($my_name, $my_version);
my %configs;

sub app_setup {
    my ($appname, $appversion, %args) = @_;
    my $help = 0;		# handled locally
    my $ident = 0;		# handled locally
    my $man = 0;		# handled locally

    my $pod2usage = sub {
        # Load Pod::Usage only if needed.
        require Pod::Usage;
        Pod::Usage->import;
        &pod2usage;
    };

    # Package name.
    $my_package = $args{package};
    # Program name and version.
    if ( defined $appname ) {
	($my_name, $my_version) = ($appname, $appversion);
    }
    else {
	($my_name, $my_version) = qw( MyProg 0.01 );
    }

    my $options =
      {
       verbose		=> 0,		# verbose processing

       ### ADD OPTIONS HERE ###

       catalog		=> undef,

       # Development options (not shown with -help).
       debug		=> 0,		# debugging
       trace		=> 0,		# trace (show process)

       # Service.
       _package		=> $my_package,
       _name		=> $my_name,
       _version		=> $my_version,
       _stdin		=> \*STDIN,
       _stdout		=> \*STDOUT,
       _stderr		=> \*STDERR,
       _argv		=> [ @ARGV ],
      };

    # Colled command line options in a hash, for they will be needed
    # later.
    my $clo = {};

    # Sorry, layout is a bit ugly...
    if ( !GetOptions
	 ($clo,

	  ### ADD OPTIONS HERE ###

	  'catalog=s',

	  # Standard options.
	  'ident'		=> \$ident,
	  'help|h|?'		=> \$help,
	  'man'			=> \$man,
	  'verbose',
	  'trace',
	  'debug',
	 ) )
    {
	# GNU convention: message to STDERR upon failure.
	$pod2usage->(2);
    }
    # GNU convention: message to STDOUT upon request.
    $pod2usage->(1) if $help;
    $pod2usage->( VERBOSE => 2 ) if $man;
    app_ident(\*STDOUT) if $ident;

    $pod2usage->(2) unless @ARGV;

    # Plug in command-line options.
    @{$options}{keys %$clo} = values %$clo;

    $options;
}

sub app_ident {
    my ($fh) = @_;
    print {$fh} ("This is ",
		 $my_package
		 ? "$my_package [$my_name $my_version]"
		 : "$my_name version $my_version",
		 "\n");
}

1;

__END__

################ Documentation ################

=head1 NAME

parse - parse MusicXML data

=head1 SYNOPSIS

parse [options] file [...]

 Options:

    --catalog=XXX	LibXML catalog.

Miscellaneous options:

    --help  -h		this message
    --man		full documentation
    --ident		show identification
    --verbose		verbose information

=head1 DESCRIPTION

This program will read the given input file(s) and parse them.

=head1 OPTIONS

=over 8

=item B<--catalog=>I<XXX>

Specifies the catalog to use. See the libxml documentation.

=item B<--help>

Prints a brief help message and exits.

=item B<--man>

Prints the manual page and exits.

=item B<--ident>

Prints program identification.

=item B<--verbose>

Provides more verbose information.

=item I<file>

The input file(s) to process.

=back

=head1 AUTHOR

Johan Vromans, C<< <jv at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-data-musicxml at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-MusicXML>. I
will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this program with the perldoc command.

    perldoc 

=head1 COPYRIGHT & LICENSE

Copyright 2016 Johan Vromans, all rights reserved.

Clone me at L<GitHub|https://github.com/sciurius/perl-Data-MusicXML>

=cut
